Cubical/Data/List/FinData.agda

```diff
@@ -34,12 +64,269 @@ lookup-tabulate : ∀ n → (^a : Fin n → A)
 lookup-tabulate (suc n) ^a i zero = ^a zero
 lookup-tabulate (suc n) ^a i (suc p) = lookup-tabulate n (^a ∘ suc) i p
 
-lookup-map : ∀ (f : A → B) (xs : List A)
-  → (p0 : Fin (length (map f xs)))
-  → (p1 : Fin (length xs))
-  → (p : PathP (λ i → Fin (length-map f xs i)) p0 p1)
-  → lookup (map f xs) p0 ≡ f (lookup xs p1)
-lookup-map f (x ∷ xs) zero zero p = refl
-lookup-map f (x ∷ xs) zero (suc p1) p = ⊥.rec (znotsP p)
-lookup-map f (x ∷ xs) (suc p0) zero p = ⊥.rec (snotzP p)
-lookup-map f (x ∷ xs) (suc p0) (suc p1) p = lookup-map f xs p0 p1 (injSucFinP p)
+lookup-tabulateT : ∀ n → (^a : Fin n → A) → ∀ k
+    → lookup (tabulate n ^a) (subst⁻ Fin (length-tabulate n ^a) k) ≡ ^a k
+lookup-tabulateT n ^a k = (sym (transportRefl _)) ∙ funExt⁻ (fromPathP (lookup-tabulate n ^a)) k
+
+permute : ∀ {n} (l : List A) → Fin n ≃ Fin (length l) → List A
+permute l e = tabulate _ (lookup l ∘ (equivFun e))
+
+length-permute : ∀ {n} (l : List A) → (e : Fin n ≃ Fin (length l)) →
+                  length (permute l e) ≡ n
+length-permute _ _ = length-tabulate _ _
+
+tabulate∘ : ∀ {n m} → (e : Fin n → A)
+               (g : Fin m → Fin n)  →
+                tabulate m (lookup (tabulate n e)
+                  ∘ subst⁻ Fin (length-tabulate _ _) ∘ g)  ≡
+                  tabulate m (e ∘ g)
+tabulate∘ {m = zero} _ _ = refl
+tabulate∘ {m = suc _} e _ =
+  cong₂ _∷_
+    ( congP (λ i → lookup-tabulate _ e i)
+     (toPathP (transportTransport⁻ (cong Fin _) _)))
+    (tabulate∘ e _)
+
+lookup-swap : ∀ (x y : A) (l : List A) → ∀ k →
+               lookup (x ∷ y ∷ l) k
+                ≡
+               lookup (y ∷ x ∷ l) (swap0and1 k)
+lookup-swap x y l zero = refl
+lookup-swap x y l one = refl
+lookup-swap x y l (suc (suc k)) = refl
+
+
+
+infix 4  _↔_
+
+record _↔_ {ℓ} {A : Type ℓ} (x y : List A) : Type ℓ where
+  constructor prm
+  field
+    F≃ : (Fin (length x) ≃ Fin (length y))
+    l≡ : ∀ k → lookup x k ≡ lookup y (equivFun F≃ k)
+
+open _↔_
+
+↔permute : ∀ {n} (l : List A) → (e : Fin n ≃ Fin (length l))
+                → l ↔ (permute l e)
+F≃ (↔permute l e) = invEquiv e ∙ₑ ≡→Fin≃ (sym (length-permute l e))
+l≡ (↔permute l e) k =
+        cong (lookup l) (sym (secEq e k))
+      ∙ sym (lookup-tabulateT _ (lookup l ∘ (fst e)) (invEq e k))
+
+
+isRefl↔ : BinaryRelation.isRefl (_↔_ {A = A})
```

Comment:
Using relation-terminology is a bit confusing, since the type you are talking about should actually be a groupoid?
Did you think about that and figured it is still better this way? Then write down your conclusion in a comment.

Reply:
Are you referring to the fact that  _↔_ is not propositional but of hLevel >= 2 ? 
(set from permutation, and possibly more from path , depending on hLevel of A) 
For Sets and Grupoids, this gives hLevel of 2.

I saw that the BinaryRelation definition does not require a relation to be propositional (it even defines the property of relation being set-valued `isSetValued`).

I used it only to make signatures more readable. Also, it let me bundle all laws into BinaryRelation.isEquivRel.

currently, I am working on groupoid truncation using this permutation relation (but proving hexagon identity is more challenging)

Reply:
Yes, that is what I was referring to. 
Just write a small comment "_↔_ is not expected to be a proposition, using Relation terminology for readability and bundling" or something like that.

Reply:
> currently, I am working on groupoid truncation using this permutation relation (but proving hexagon identity is more challenging)

Not sure I understand. Sounds interesting - if you like to discuss it, just make a draft PR and I'll have a look.

Reply:
You can always take the propositional truncation of your relation and make it prop-valued, but when you take the set-quotient you end up with the same thing either way. 

That's a little lemma called [truncRelIso](https://github.com/agda/cubical/blob/master/Cubical/HITs/SetQuotients/Properties.agda#L282) and I always took it as justification for quotienting by non-prop-valued relations. 

As long as your relation is an equivalence relation in the sense above, it gives you a nice characterization of the quotients path type: [isEquivRel→TruncIso](https://github.com/agda/cubical/blob/master/Cubical/HITs/SetQuotients/Properties.agda#L295)

Reply:
Ah, I thought that's true, but I didn't know about the lemma. I guess we use the terminology somewhere else in the same way and withdraw my request for a comment.